mbuild/compound.py

```diff
@@ -1138,7 +1138,159 @@ def _kick(self):
             particle.pos += (np.random.rand(3,) - 0.5) / 100
         self._update_port_locations(xyz_init)
 
-    def energy_minimization(self, steps=2500, algorithm='cg',
+    def energy_minimization(self, forcefield='UFF',steps=1000, ):
+        """Perform an energy minimization on a Compound
+
+        Utilizes Open Babel (http://openbabel.org/docs/dev/) to perform an
+        energy minimization/geometry optimization on a Compound by applying
+        a generic force field or OpenMM (http://openmm.org/) to energy minimize
+        using a user-defined forcefield xml.
+
+        This function is primarily intended to be used on smaller components,
+        with sizes on the order of 10's to 100's of particles, as the energy
+        minimization scales poorly with the number of particles.
+
+        Parameters
+        ----------
+        steps : int, optionl, default=1000
+            The number of optimization iterations
+        forcefield : str, optional, default='UFF'
+            The generic force field to apply to the Compound for minimization.
+            Valid options are 'MMFF94', 'MMFF94s', ''UFF', 'GAFF', and 'Ghemical'.
+            Please refer to the Open Babel documentation (http://open-babel.
+            readthedocs.io/en/latest/Forcefields/Overview.html) when considering
+            your choice of force field.
+        **kwargs : keyword arguments
+            Key word arguments passed to _minimize_energy_openbabel or 
+            _minimize_energy_openmm
+
+        """
+        tmp_dir = tempfile.mkdtemp()
+        original = clone(self)
+        self._kick()
+        self.save(os.path.join(tmp_dir,'un-minimized.mol2'))
+        extension = os.path.splitext(forcefield)[-1]
+        if extension == '.xml':
+            self._minimize_energy_openmm(tmp_dir, forcefield=forcefield,
+                    steps=steps)
+
+        else:
+            self._minimize_energy_openbabel(tmp_dir, forcefield=forcefield,
+                    steps=steps)
+
+
+        self.update_coordinates(os.path.join(tmp_dir, 'minimized.pdb'))
```

Comment:
We've had issues in the past using PDB as the intermediate file format for `update_coordinates`. Those issues should have been fixed in #364, but we'll want to make sure we have plenty of tests to make sure nothing unexpected is happening here.

Reply:
OpenMM can only report DCD, PDB, PDBx/mmCIF. For the sake of re-using code, I tried to keep the output of openbabel and openmm as PDB, and then use the same line for `self.update_coordinates`. If this is not worth the potential extra troubles, I can keep `_energy_minimize_openbabel` utilizing mol2s, but have `_energy_minimize_openmm` utilize pdb.

Reply:
Not sure if this is directly useful but there are other custom reports which we already import via our dependencies:

http://mdtraj.org/development/api/reporters.html
https://parmed.github.io/ParmEd/html/api/parmed/parmed.openmm.reporters.html